protocols/kad/src/behaviour.rs

```diff
@@ -1449,6 +1449,34 @@ where
         self.connected_peers.insert(peer.clone());
     }
 
+    fn inject_address_change(
+        &mut self,
+        peer: &PeerId,
+        _: &ConnectionId,
+        old: &ConnectedPoint,
+        new: &ConnectedPoint
+    ) {
+        // Only addresses of outbound connections are inserted into the routing
+        // table in `inject_connection_established`. Thereby only address
+        // changes of outbound connections require updating.
+        let new_address = match new {
+            ConnectedPoint::Dialer { address } => Some(address.clone()),
+            ConnectedPoint::Listener { .. } => return,
```

Comment:
The reason `inject_connection_established` only directly tries to attempt insertion into the routing table for `Dialer` addresses, is that these are known to be "routable" / "dialable". In contrast, the listener address only _may_ be dialable and we emit an `UnroutablePeer` event. User code may subsequently add this address to the routing table if it is deemed dialable. Whether the `Listener` address is routable / dialable in general depends on the transport protocol. With TCP it isn't, as it contains a connection-specific ephemeral client port number but in general it may be (e.g. with QUIC?). So I think the safer approach here for `inject_address_change` is: If the `old` address is in the routing table for `peer`, replace it with `new`, regardless of whether it is a `Dialer` or `Listener` address. It is probably also fine to always insert `new` even if `old` is not present (but only if the peer is already in the routing table, of course).